### PR TITLE
[WIP] Finish REST API

### PR DESCRIPTION
**Pull request by kashizui**
Contributes to #252 

REST-ified
- [x] gls
- [x] gnew
- [x] grm
- [x] ginfo
- [x] uadd
- [x] urm
- [x] search (Sunday 5/1)
- [x] info, metadata only (Sunday 5/1)
- [x] edit (Monday 5/2)
- [x] rm (Monday 5/2)
- [x] perm (Monday 5/2)
- [x] chown (Monday 5/2)
- [x] run (Tuesday 5/3)
- [x] make (Wednesday 5/4)
- [x] mimic, shadow-only (Wednesday 5/4)
- [x] macro, shadow-only (Thursday 5/5)
- [x] wait (Friday 5/6)
- [x] download (Saturday 5/7)
- [x] cat (Saturday 5/7)
- [x] info, with contents (Sunday 5/8)
- [x] [web bundle details page](https://github.com/codalab/codalab-worksheets/pull/162) (Monday 5/9)
- [ ] upload (Thursday 5/12)
- [ ] ls (Friday 5/13)
- [ ] work
- [ ] print
- [ ] new
- [ ] wedit
- [ ] wls
- [ ] wrm
- [ ] add
- [ ] detach
- [ ] wadd
- [ ] wperm
- [ ] kill
- [ ] write
- [ ] logout
- [ ] all tests passing

### CODE REVIEW

File: codalab/rest/json_api_client.py [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+                    sys.exc_info()[2]
+            except (urllib2.URLError, httplib.HTTPException, socket.error) as e:
+                raise JsonApiException, \
+                    JsonApiException(message + ': ' + str(e), False), \
+                    sys.exc_info()[2]
+        return wrapper
+    return decorator
+
+
+def unpack_json(raw_json):
```
Comment (percyliang) [inline codalab/rest/json_api_client.py]: I was under the impression that marshmallow did this for you...

File: codalab/rest/groups.py [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+        local.model.update_user_in_group(membership['user_id'], membership['group_uuid'], membership['is_admin'])
+    else:
+        membership = membership_updates
+        membership = local.model.add_user_in_group(membership['user_id'], membership['group_uuid'], membership['is_admin'])
+
+    return GroupMembershipSchema().dump(membership).data
+
+
+@delete('/group-memberships/<membership_id>', apply=AuthenticatedPlugin())
+def delete_group_membership(membership_id):
```
Comment (percyliang) [inline codalab/rest/groups.py]: Is delete with a single membership_id standard?

I thought that with the JSON API, you would be sending some sort of JSON message to delete.

What if you wanted to delete multiple memberships in batch?

Don't you also need to check that the user has permissions to delete this?

Comment (kashizui) [inline codalab/rest/groups.py]: > Is delete with a single membership_id standard?
> I thought that with the JSON API, you would be sending some sort of JSON message to delete.
> What if you wanted to delete multiple memberships in batch?

Yes but we can support bulk deletes with requests like this:

```
DELETE /articles HTTP/1.1
Content-Type: application/vnd.api+json; ext=bulk
Accept: application/vnd.api+json; ext=bulk

{
  "data": [
    { "type": "articles", "id": "1" },
    { "type": "articles", "id": "2" }
  ]
}
```

Perhaps we can simplify and only allow bulk deletes?

> Don't you also need to check that the user has permissions to delete this?

The `need_admin=True` forces a permission check--it's not very clear though to be sure, I just copied the `get_group_info` function from `LocalBundleClient`.

File: codalab/rest/json_api_client.py (hunk 1 above)
Comment (kashizui) [inline codalab/rest/json_api_client.py]: I experimented with deserializing on the client side with marshmallow, but I was struggling to think of a good way to represent, for example, the username of the owner on the client side. If we deserialize back into the original group_info, we get:

```
{
    "memberships": [3, 6, 7],
    "name": "ok-group",
    "owner_id": "6f2e70e0fed2487185fe0f21fffde98f",
    "uuid": "0x4612d33b6709410895b2bee028e05880"
}
```

We also get the owner object and each of the membership objects included as part of the compound response, but in order to look up the owner's username, we would have to manually search through these included objects to find the user matching the owner_id. This search procedure also would not generalize to other resource types (e.g. searching through bundle permissions).

I instead opted to make the CLI/client side ignore any prior knowledge about the structure of the internal DB dicts, and instead just use these client-side JSON API attributes as is, and do a little bit of processing work to wire up the relationships. So instead we get this:

```
{
    "id": "0x4612d33b6709410895b2bee028e05880",
    "memberships": [
        {
            "group": {
                "id": "0x4612d33b6709410895b2bee028e05880"
            },
            "id": 3,
            "is_admin": true,
            "user": {
                "affiliation": "SCOr",
                "date_joined": "Mon Mar 28 22:49:17 2016",
                "email": "sckoo@cs.stanford.edu",
                "first_name": "\u9756\u5d34",
                "id": "6f2e70e0fed2487185fe0f21fffde98f",
                "last_name": "St\u00e9ph\u00e5n",
                "url": "http://sckoo.net/though",
                "user_name": "sckoo"
            }
        },
        {
            "group": {
                "id": "0x4612d33b6709410895b2bee028e05880"
            },
            "id": 6,
            "is_admin": true,
            "user": {
                "affiliation": null,
                "date_joined": "Fri Apr  8 20:53:24 2016",
                "email": "newuser@gmail.com",
                "first_name": null,
                "id": "fbeb0aebabb84da8b844ac9fba4fe95a",
                "last_name": null,
                "url": null,
                "user_name": "newuser"
            }
        },
        {
            "group": {
                "id": "0x4612d33b6709410895b2bee028e05880"
            },
            "id": 7,
            "is_admin": true,
            "user": {
                "affiliation": null,
                "date_joined": "Fri Apr  8 21:03:21 2016",
                "email": "newuser2@gmail.com",
                "first_name": null,
                "id": "9af28fa6a8e140e6a50561f3fcdb040e",
                "last_name": null,
                "url": null,
                "user_name": "newuser2"
            }
        }
    ],
    "name": "ok-group",
    "owner": {
        "affiliation": "SCOr",
        "date_joined": "Mon Mar 28 22:49:17 2016",
        "email": "sckoo@cs.stanford.edu",
        "first_name": "\u9756\u5d34",
        "id": "6f2e70e0fed2487185fe0f21fffde98f",
        "last_name": "St\u00e9ph\u00e5n",
        "url": "http://sckoo.net/though",
        "user_name": "sckoo"
    }
}
```

Now we can directly get the owner's username with `group['owner']['user_name']`. And we also don't need to choose the right deserializer for each of the objects in response, and we don't have to assume anything about how the objects are internally represented in the DB model. The `unpack_json` operation is the same for all responses.

Now I'm defining the function of marshmallow as: transforming db dicts into the JSON API format when formulating responses, and **validating** input data from JSON API requests and transforming them back into directly usable db dicts. I was able to completely factor out the Group ORMObject, and I'm confident I can factor out the Bundle ORMObjects with this as well. Marshmallow is in this case `not` being used on the client side.

Comment (percyliang) [inline codalab/rest/json_api_client.py]: I think that makes sense.

File: codalab/rest/groups.py (hunk 2 above)
Comment (percyliang) [inline codalab/rest/groups.py]: What code actually gets invoked when you make your bulk delete request?  Simplifying sounds good.

Regarding `need_admin=True`, but one shouldn't need admin access to modify one's own permissions...(perhaps I'm misinterpreting something).

Comment (kashizui) [inline codalab/rest/groups.py]: I haven't written that code yet, but I believe bulk delete code exists for bundles if not for groups.

What modifications to one's own permissions can one do without admin access?

File: codalab/rest/json_api_client.py (hunk 1 above)
Comment (kashizui) [inline codalab/rest/json_api_client.py]: If there's anything that seems funny though please let me know lest I screw something up! :)

File: codalab/rest/groups.py (hunk 2 above)
Comment (percyliang) [inline codalab/rest/groups.py]: Never mind, you mean admin of the group.

**Comment by kashizui** [comment]
I've decided that it's been causing too much trouble to delete existing/redundant code as I go, since the dependencies are still too intricate not to break thing unintentionally. I'll go back and delete everything once the entire REST API is ready, but in the meanwhile please bear with me, there will be a lot of additional code that is mostly adapted from existing code.

**Comment by kashizui** [comment]
@percyliang should I finish out the REST API on this one PR, or try to separate into two+ PRs?

**Comment by kashizui** [comment]
@klopyrev I've cleaned up the PR, let me know if there's anything else that might be unclear/unnecessary

File: codalab/client/rest_client.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    client_error is False, the failure is caused by a server-side error and
+    can be retried.
+    """
+
+    def __init__(self, message, client_error):
+        super(RestClientException, self).__init__(message)
+        self.client_error = client_error
+
+
+class RestClient(object):
```
Comment (klopyrev) [inline codalab/client/rest_client.py:21]: Put this file into a separate pull request.

File: codalab/common.py [hunk 4]
```diff
@@ -57,12 +58,20 @@ class PermissionError(UsageError):
     necessary permissions. Similar to HTTP status 403.
     """
 
+
+class InternalError(StandardError):
```
Comment (klopyrev) [inline codalab/common.py]: PreconditionViolation already does this. No need for this extra class.

File: codalab/client/rest_client.py (hunk 3 above)
Comment (kashizui) [inline codalab/client/rest_client.py:21]: As in, keep my separate JsonApiClient implementation for now in this PR? (I was hoping to use the upload_with_chunked_encoding method, so I should just make a copy of it for now?)

File: codalab/rest/bundle.py [hunk 5]
```diff
@@ -1,15 +1,546 @@
...
+
+    check_bundles_have_read_permission(local.model, request.user, [uuid])
+    return {
+        'data': local.download_manager.get_target_info(uuid, path, depth)
+    }
+
+
+@get('/bundles/<uuid:re:%s>/contents/blob/' % spec_util.UUID_STR, apply=AuthenticatedPlugin())
+@get('/bundles/<uuid:re:%s>/contents/blob/<path:path>' % spec_util.UUID_STR, apply=AuthenticatedPlugin())
+def fetch_contents_blob(uuid, path=''):
```
Comment (klopyrev) [inline codalab/rest/bundle.py:543]: Put this change as well as all the places that use this API into a separate pull request.

File: codalab/rest/bundle.py [hunk 6]
```diff
@@ -1,15 +1,546 @@
...
+    if depth < 0:
+        abort(httplib.BAD_REQUEST, "Depth must be at least 0")
+
+    check_bundles_have_read_permission(local.model, request.user, [uuid])
+    return {
+        'data': local.download_manager.get_target_info(uuid, path, depth)
+    }
+
+
+@get('/bundles/<uuid:re:%s>/contents/blob/' % spec_util.UUID_STR, apply=AuthenticatedPlugin())
```
Comment (klopyrev) [inline codalab/rest/bundle.py:541]: The worker calls this method and you haven't updated it to rename the API. Do you really need to rename the API? If you submit this after I launch the new worker system renaming it would be bad.

Comment (kashizui) [inline codalab/rest/bundle.py:541]: Hm yeah that's why I asked you about this yesterday. What are your thoughts? We could 1) keep this as the singular "bundle" and just deal with it being different from the other, 2) change all the bundle methods to be singular, 3) change ALL the methods to be singular, 4) do this renaming in a separate PR that is merged ASAP.

File: codalab/rest/bundle.py (hunk 5 above)
Comment (kashizui) [inline codalab/rest/bundle.py:543]: Are you referring to the route name change, or the code for handling the extra query parameters? The latter is required to remove the CLI's dependence on LocalBundleClient.

**Comment by kashizui** [comment]
Split out into multiple pull-requests:
#431 
#432 
#433 
#434 
and more

File: codalab/rest/bundle.py (hunk 6 above)
Comment (kashizui) [inline codalab/rest/bundle.py:541]: @klopyrev what were your thoughts on this? I'd like to do my best to keep you unblocked for deploying your worker system.
